security_monkey/scheduler.py

```diff
@@ -38,24 +38,61 @@ def run_change_reporter(account_names, interval=None):
 
 def find_changes(accounts, monitor_names, debug=True):
     """
-        Runs the watcher and stores the result, reaudits all types to account
+        Runs the watcher and stores the result, re-audits all types to account
         for downstream dependencies.
     """
     for account_name in accounts:
         monitors = get_monitors(account_name, monitor_names, debug)
         for mon in monitors:
             cw = mon.watcher
-            (items, exception_map) = cw.slurp()
-            cw.find_changes(current=items, exception_map=exception_map)
-            cw.save()
+            if mon.batch_support:
+                audit_items = []
+
+                # Fetch the full list of items that we need to obtain:
+                _, exception_map = cw.slurp_list()
+                if len(exception_map) > 0:
+                    app.logger.error("Exceptions have caused nothing to be fetched for {technology}/{account}..."
+                                     " CANNOT CONTINUE FOR THIS WATCHER!".format(technology=cw.i_am_plural,
+                                                                                 account=account_name))
+                    continue
+
+                while not cw.done_slurping:
+                    app.logger.debug("Fetching a batch of {batch} items for {technology}/{account}.".format(
+                        batch=cw.batched_size, technology=cw.i_am_plural, account=account_name
+                    ))
+                    (items, exception_map) = cw.slurp()
+
+                    audit_items.extend(cw.find_changes(current=items, exception_map=exception_map))
+
+                # Delete the items that no longer exist:
+                app.logger.debug("Deleting all items for {technology}/{account} that no longer exist.".format(
+                    technology=cw.i_am_plural, account=account_name
+                ))
+                cw.find_deleted_batch(account_name)
+                _audit_specific_changes(mon, audit_items, False, debug)
```

Comment:
Need to test what exceptions do to the flow -- does it contain it or does it just kill the whole thing?